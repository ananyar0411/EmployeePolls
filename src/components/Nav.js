import { Link } from "react-router-dom";
import React from "react";

import { connect, useDispatch } from "react-redux";
import "../../src/index.css";
import { setAuthedUser } from "../actions/authedUser";

const Nav = ({ authedUser, users }) => {
    const dispatch = useDispatch();

    const user = users.find((u) => u.id === authedUser);

    function handleLogout() {
        dispatch(setAuthedUser(null));
        window.location.href = "/";
    }

    return (
        <nav className="navbar">
            <ul>
                {authedUser && (
                    <>
                        <li>
                            <Link to="/home">Home</Link>
                        </li>
                        <li>
                            <Link to="/add">New Poll</Link>
                        </li>
                        <li>
                            <Link to="/leaderboard">Leaderboard</Link>
                        </li>
                    </>
                )}
                <li>
                    {authedUser ? (
                        <Link onClick={handleLogout}>Logout</Link>
                    ) : (
                        <Link to="/">Login</Link>
                    )}
                </li>
            </ul>
            {user && (
                <div className="user">
                    <img src={user.avatarURL} alt={`Avatar of user`} />
                    <h3>{user.id}</h3>
                </div>
            )}
        </nav>
    );
};
const mapStateToProps = ({ authedUser, users }) => ({
    users: Object.values(users),
    authedUser,
});

export default connect(mapStateToProps)(Nav);
