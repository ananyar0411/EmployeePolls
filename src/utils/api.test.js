import {
    _getUsers,
    _getQuestions,
    _saveQuestionAnswer,
    _saveQuestion,
} from "./api";
import { fireEvent, render } from "@testing-library/react";
import React from "react";
import { createStore } from "redux";
import reducer from "../reducers";
import middleware from "../middleware";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";

import { BrowserRouter as Router } from "react-router-dom";
import Nav from "../components/Nav";

console.log(reducer);
const store = createStore(reducer, middleware);

describe("getusers", () => {
    it("get the users", async () => {
        var l = 4;
        var result = await _getUsers();

        expect(Object.keys(result).length).toEqual(l);
    });
});

describe("_getQuestions", () => {
    it("get the questions", async () => {
        var l = 6;
        var result = await _getQuestions();

        expect(Object.keys(result).length).toEqual(l);
    });
});

describe("_saveQuestion", () => {
    it("Saved question success", async () => {
        const question = {
            optionOneText: "one",
            optionTwoText: "two",
            author: "sarahedo",
        };

        var result = (await _saveQuestion(question)).formattedQuestion;
        expect(result.author).toEqual("sarahedo");
        expect(result.optionOne.text).toEqual("one");
        expect(result.optionTwo.text).toEqual("two");
    });

    it("failed to save question", async () => {
        const emptyObject = {};
        await expect(_saveQuestion(emptyObject)).rejects.toEqual(
            "Please provide optionOneText, optionTwoText, and author"
        );
    });

    it("failed to save question on no user logged in", async () => {
        const question = {
            optionOneText: "one",
            optionTwoText: "two",
            author: null,
        };

        await expect(_saveQuestion(question)).rejects.toEqual(
            "Please provide optionOneText, optionTwoText, and author"
        );
    });

    it("failed to save question when user doesn't exist", async () => {
        const question = {
            optionOneText: "one",
            optionTwoText: "two",
            author: "johndoe",
        };

        await expect(_saveQuestion(question)).rejects.toEqual(
            "User doesn't exist"
        );
    });
});

describe("_saveQuestionAnswer", () => {
    it("successful save question answer", async () => {
        const testObj = {
            authedUser: "sarahedo",
            qid: "8xf0y6ziyjabvozdd253nd",
            answer: "optionOne",
        };
        var { result } = await _saveQuestionAnswer(testObj);
        expect(result).toEqual(true);
    });

    it("failed to save empty question answer", async () => {
        const emptyObject = {};
        await expect(_saveQuestionAnswer(emptyObject)).rejects.toEqual(
            "Please provide authedUser, qid, and answer"
        );
    });

    it("failed to save on unauthed user", async () => {
        const testObj = {
            authedUser: null,
            qid: "8xf0y6ziyjabvozdd253nd",
            answer: "optionOne",
        };
        await expect(_saveQuestionAnswer(testObj)).rejects.toEqual(
            "Please provide authedUser, qid, and answer"
        );
    });

    it("failed to save answer for a user which is not defined", async () => {
        const testObj = {
            authedUser: "Jane Doe",
            qid: "8xf0y6ziyjabvozdd253nd",
            answer: "optionOne",
        };
        await expect(_saveQuestionAnswer(testObj)).rejects.toEqual(
            "user does not exist"
        );
    });
});

describe("render navbar correctly", () => {
    it("Matches DOM Snapshot", () => {
        const tree = renderer.create(
            <Provider store={store}>
                <Router>
                    <Nav />
                </Router>
            </Provider>
        );
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it("redirects user to login", async () => {
        const tree = render(
            <Provider store={store}>
                <Router>
                    <Nav />
                </Router>
            </Provider>
        );
        const select = await tree.findByText("Login");
        fireEvent.click(select);
        expect(window.location.pathname).toEqual("/");
    });
});
