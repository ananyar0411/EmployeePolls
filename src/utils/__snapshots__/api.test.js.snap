// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render navbar correctly Matches DOM Snapshot 1`] = `
<nav
  className="navbar"
>
  <ul>
    <li>
      <a
        href="/"
        onClick={[Function]}
      >
        Login
      </a>
    </li>
  </ul>
</nav>
`;
